File: sections/14_core_patterns_mixed_contest/problems/o_minimum_parenthesis_depth/solution.cpp
#include <bits/stdc++.h>
using namespace std;

bool feasible(const string &s, int limit) {
    int low = 0, high = 0;
    for (int i = 0; i < (int)s.size(); ++i) {
        if (s[i] == '(') ++low, ++high;
        else if (s[i] == ')') --low, --high;
        else --low, ++high;
        int parity = (i + 1) & 1;
        if (low < 0) low = parity;
        high = min(high, limit);
        if ((high & 1) != parity) --high;
        if (low > high) return false;
    }
    return low == 0;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s;
    cin >> s;
    if (s.size() % 2 || !feasible(s, (int)s.size())) {
        cout << -1 << '\n';
        return 0;
    }
    int low = 1, high = s.size();
    while (low < high) {
        int middle = low + (high - low) / 2;
        if (feasible(s, middle)) high = middle;
        else low = middle + 1;
    }
    cout << low << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/o_minimum_parenthesis_depth/solution.py
import sys


def feasible(s: str, limit: int) -> bool:
    low = high = 0
    for prefix, char in enumerate(s, 1):
        if char == "(":
            low += 1
            high += 1
        elif char == ")":
            low -= 1
            high -= 1
        else:
            low -= 1
            high += 1
        parity = prefix & 1
        if low < 0:
            low = parity
        high = min(high, limit)
        if high & 1 != parity:
            high -= 1
        if low > high:
            return False
    return low == 0


def main() -> None:
    s = sys.stdin.buffer.readline().strip().decode()
    if len(s) % 2 or not feasible(s, len(s)):
        print(-1)
        return
    low, high = 1, len(s)
    while low < high:
        middle = (low + high) // 2
        if feasible(s, middle):
            high = middle
        else:
            low = middle + 1
    print(low)


if __name__ == "__main__":
    main()
